### DISCUSSION TITLE
CSLA 6 way to LazyGetProperty?

### DISCUSSION THREAD

**Discussion by michaelcsikos** [MarimerLLC/csla#2829] [Questions]
What's the new way to `LazyGetProperty`? I can't see anything in the samples.

**Comment by rockfordlhotka** [comment]
```c#
      get => LazyGetProperty<PersonEdit>(ChildProperty, () 
        => ApplicationContext.GetRequiredService<IDataPortal<<PersonEdit>>().Fetch(123));
```

or

```c#
      get => LazyGetPropertyAsync<PersonEdit>(ChildProperty, 
        ApplicationContext.GetRequiredService<IDataPortal<<PersonEdit>>().FetchAsync(123));
```

**Reply by michaelcsikos** [reply]
What about for a child business object property that is not part a collection? e.g. `AddressEdit PostalAddress`,

Edit: OK, I see this is specifically for child types. It looked wrong calling `Create` or `Fetch` instead of the `IChildDataPortal` versions.

**Reply by rockfordlhotka** [reply]
Normally you'd use an intermediate command object to act as a getter for the child.

```c#
[Serializable]
public class PersonGetter : ReadOnlyBase<PersonGetter>
{
    // implement Person property

    [Fetch]
    private void Fetch(int id, [Inject] IChildDataPortal<Person> portal)
    {
        Person = portal.FetchChild(id);
    }
}
```

**Reply by rockfordlhotka** [reply]
Then

```c#
      get => LazyGetPropertyAsync<PersonGetter>(ChildProperty, GetDataPortal<PersonGetter>().FetchAsync(123).Person);
```

**Reply by michaelcsikos** [reply]
In this case, I'm only creating the child object on demand. When the root object is fetched, any children will be loaded immediately. Is `RelationshipTypes.LazyLoad` still required? There's no `RegisterProperty<P>(string propertyName, string friendlyName, RelationshipTypes relationship)` overload.

**Reply by rockfordlhotka** [reply]
Yes, it is required. I think that overload was ambiguous, so you need to choose one of the other overloads.

**Reply by michaelcsikos** [reply]
OK, so what does `P defaultValue` mean in the context of a  lazy-loaded property? Seems redundant; do we just pass `null`?

**Reply by rockfordlhotka** [reply]
Either `null` or `default(P)` - but I'd think `null`.

**Reply by michaelcsikos** [reply]
In a parent's `Fetch` is there anything wrong with:

```cs
var childPortal = (IChildDataPortal<Child>)GetDataPortal<Child>();

Child           = childPortal.FetchChild(Id);
```

**Reply by michaelcsikos** [reply]
Ah, I see I should just `[Inject] IChildDataPortal<Child> childPortal` in the parent's `Fetch` method. It's kind of magical, isn't it.

**Reply by TheCakeMonster** [reply]
@michaelcsikos Did this work for you? I'm trying this in the unit tests and it appears that FieldManager.FieldExists always returns true, so the delegate used for loading of the data is never called.

**Reply by TheCakeMonster** [reply]
Ah, wait, I get it. RelationshipTypes.LazyLoad is required as part of the RegisterProperty call. I should have read further down the thread. Sorry!

**Reply by TheCakeMonster** [reply]
It looks like this answer might need updating now, as I think the GetDataPortal method has been removed.

**Reply by Chicagoan2016** [reply]
@TheCakeMonster I thought we made the RelationshipTypes enum obsolete?

**Reply by rockfordlhotka** [reply]
Not the enum, but some members. The LazyLoad member (I think) is the only one still used.

**Reply by KC7465128305** [reply]
Hi Rocky,

I just upgraded Csla from 5.5.3 to 6.1.0 and the FieldManager.FieldExists always returns true in a collection property causing the property to return null which breaks my code.

Here is my existing old code in Csla 5.5.3. What is the equivalent in Csla 6.1.0?
public static readonly PropertyInfo<Operands> OperandsProperty = RegisterProperty<Operands>(p => p.Operands);
internal Operands Operands
{
       get
       {
             if (!(FieldManager.FieldExists(OperandsProperty)))
                 LoadProperty<Operands>(OperandsProperty,
                                                this.IsNew
                                                    ? Operands.NewOperands()
                                                    : Operands.GetOperands(this));

                return GetProperty<Operands>(OperandsProperty);
        }
}

Thanks very much if you can help!

**Reply by rockfordlhotka** [reply]
When you declare the property via `RegisterProperty` are you setting the relationship type?

`RelationshipTypes.LazyLoad` is required.

**Comment by KC7465128305** [comment]
Hi Rocky,

That fixes it.  Thanks.

After I upgraded Csla from 5.5.3 to 6.1.0, I noticed there is a noticeable slowness in the app that I didn't notice when I upgraded Csla from 4.X to 5.X.  My code in 5.5.3 is still the old-style DataPortal_XYZ and I only use DI in 6.1.0. and I suspect this is the reason it is slower.  I read that Csla 5.X do a lot of things Csla 4.X didn't do and this account for its slowness.  I only use Csla for one thing, and that is N Level Undo.  I wonder if there is a way you have an enum so users of Csla can pick and choose what functions they want to enable and what functions they can disable. This way a user can save a lot of overhead for functions they don't use anyway.

If this is possible, it will be great.  Thanks!

**Comment by ashishsinha24** [comment]
I'm encountering difficulties configuring a straightforward scenario. Imagine I have a database with two tables: "Person" containing Id , Name and CityId, and "Location" with CityId, CityName and State.

I've successfully created a ReadOnlyListBase for persons, but when attempting to include the location information (which is in a readonlybase LocationInfo) in PersonInfo, various errors arise, such as:

ChildDataportal failure.
Stack overflow.

My objective is to display the Id, PersonName, City, and State in a grid. While I can successfully retrieve the PersonId and Name, issues arise when trying to integrate location details.

**Comment by kcabral817** [comment]
The stack overflow can occur if there is a collection of persons that have location.  If you're using EF Core, their maybe a virtual member that points back to the Person.  This is where I've experienced a stack overflow because it'll be self referencing for ever.  But at this point, it's purely conjecture on my part, because I would need to see the class structure and how are you mapping between the EF core (if used) and the CSLA class?  If you're using Automapper, I'd start there.

But in any case, I would need to see the class structures to assit.


Have a nice day.

Kevin

https://rb.gy/syiuf


________________________________
